src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -501,9 +515,39 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
             return this.GetTriviaData(tokenData1, tokenData2).SecondTokenIsFirstTokenOnLine;
         }
 
+        /// <summary>
+        /// Gets approximated index bounds for a token with specified position, which will definitely contain
+        /// the token, if it is present in the stream.
+        /// </summary>
+        private void GetTokenIndexBounds(int position, out int lowerBound, out int upperBound)
+        {
+            float avgLength = this.AverageTokenLength;
+            Debug.Assert(avgLength > 0);
+            // make an initial guess about the token position
+            lowerBound = upperBound = Bounded((int)((position - _tokens[0].FullSpan.Start) / avgLength), min: 0, max: _tokens.Length - 1);
+
+            // decrease lower bound until the desired position is strictly after it
+            while (_tokens[lowerBound].FullSpan.Start >= position && lowerBound > 0)
+            {
+                int estimatedError = (int)((_tokens[lowerBound].FullSpan.Start - position) / avgLength);
+                lowerBound = Math.Max(0, lowerBound - Math.Max(1, estimatedError));
+            }
+
+            // increase upper bound until the desired position is strictly before it
+            while (_tokens[upperBound].FullSpan.Start <= position && upperBound < _tokens.Length - 1)
+            {
+                int estimatedError = (int)((position - _tokens[upperBound].FullSpan.Start) / avgLength);
+                upperBound = Math.Min(_tokens.Length - 1, upperBound + Math.Max(1, estimatedError));
+            }
+        }
+
+        private static int Bounded(int value, int min, int max)
+            => Math.Min(max, Math.Max(value, min));
```

Comment:
i've written this same piece of code like 10 times in my life, and it still always trips me up :)